### PR TITLE
New module for R:D:capabilities

### PR DESCRIPTION
**Pull request by oernst9**
* created new module for capabilities according to AARC specification. i.e. specification of resource and optional actions.

* created tests for the checkAttributeSyntax method for the module

### CODE REVIEW

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_resource_attribute_def_def_capabilities.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+		"+,)*(" + actionAllowedCharacters + "+)))?");
+
+
+	@Override
+	public void checkAttributeSyntax(PerunSessionImpl perunSession, Resource resource, Attribute attribute) throws InternalErrorException, WrongAttributeValueException {
+		if (attribute.getValue() == null) return;
+		List<String> values = attribute.valueAsList();
+		for (String value : values) {
+			if (!pattern.matcher(value).matches()) {
+				throw new WrongAttributeValueException(attribute, resource.getName() + " has attribute whose value is not valid");
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_resource_attribute_def_def_capabilities.java]: Add there an example of valid value
`res:RESOURCE[:CHILD-RESOURCE1][:CHILD-RESOURCE2]...[:act:ACTION[,ACTION]...]`

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_resource_attribute_def_def_capabilities.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+ * Example2: value=“res:RESOURCE1:CHILD-RESOURCE1:CHILD-RESOURCE2”
+ * Example3: value=“res:=$:(:)”
+ *
+ * @author Ondrej Ernst
+ */
+public class urn_perun_resource_attribute_def_def_capabilities extends ResourceAttributesModuleAbstract implements ResourceAttributesModuleImplApi {
+
+	private static final String resourceAllowedCharacters = "[a-zA-Z0-9()+,\\-.=@;$_!*']";
+	private static final String actionAllowedCharacters = "[a-zA-Z0-9()+\\-.:=@;$_!*']";
+	private static final Pattern pattern = Pattern.compile("res((?!(:act$|:act:|.*:,|.*,:|.*::|.*:$))(:" + // what the whole string cannot contain
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_resource_attribute_def_def_capabilities.java]: You can separate whole negative regex look-ahead logic to variable as:
```
private static final String notAllowedParts = "?!(:act$|:act:|.*:,|.*,:|.*::|.*:$)";
```
And then change the regex as:
```
Pattern.compile(
"res(" + notAllowedParts + ")(:" + resourceAllowedCharacters + "+))+(:act:((" + actionAllowedCharacters + "+,)*(" + actionAllowedCharacters + "+)))?"
);
```
to be easier human readable.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_resource_attribute_def_def_capabilities.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+import java.util.List;
+import java.util.regex.Pattern;
+
+/**
+ * Module for capabilities according to AARC specification. i.e. specification of resource and optional actions.
+ * Syntax:
+ * value=“res:RESOURCE[:CHILD-RESOURCE1][:CHILD-RESOURCE2]...[:act:ACTION[,ACTION]...]”
+ * Example1: value=“res:RESOURCE1:CHILD-RESOURCE1:CHILD-RESOURCE2:act:ACTION1,ACTION2,ACTION3”
+ * Example2: value=“res:RESOURCE1:CHILD-RESOURCE1:CHILD-RESOURCE2”
+ * Example3: value=“res:=$:(:)”
```
Comment (licehammer) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_resource_attribute_def_def_capabilities.java]: Delete this line please. Although the value it's valid, it doesn't works well as a example - it just makes people wondering what does it mean.
If you don't like my suggestion, you can mark it as a example of value, which doesn't make much sense, but it's syntactically correct.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_resource_attribute_def_def_capabilities.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+ * @author Ondrej Ernst
+ */
+public class urn_perun_resource_attribute_def_def_capabilities extends ResourceAttributesModuleAbstract implements ResourceAttributesModuleImplApi {
+
+	private static final String resourceAllowedCharacters = "[\\-a-zA-Z0-9()+,.=@;$_!*']";
+	private static final String actionAllowedCharacters = "[\\-a-zA-Z0-9()+.:=@;$_!*']";
+	private static final String notAllowedParts = "?!(:act$|:act:|.*:,|.*,:|.*::|.*:$)";
+	private static final Pattern pattern = Pattern.compile("res((" + notAllowedParts + ")(:" +
+		resourceAllowedCharacters + "+))+(:act:((" + actionAllowedCharacters +
+		"+,)*(" + actionAllowedCharacters + "+)))?");
```
Comment (licehammer) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_resource_attribute_def_def_capabilities.java]: According of Java naming conventions constants should be written in uppercase with underscores.

**Pull request merged**